registry/txt.go

```diff
@@ -129,13 +129,31 @@ func (im *TXTRegistry) ApplyChanges(ctx context.Context, changes *plan.Changes)
 		UpdateOld: filterOwnedRecords(im.ownerID, changes.UpdateOld),
 		Delete:    filterOwnedRecords(im.ownerID, changes.Delete),
 	}
+
+	// Retain ProviderSpecific properties from existing TXT record if present
+	providerSpecificPropertyMap := map[string]endpoint.ProviderSpecific{}
+	records, err := im.provider.Records()
```

Comment:
I'm not so sure that this is a good idea. This will make the provider implementation retrieve all the records again, which might be an expensive operation.

Reply:
The problem here is that with Designate and txt registry we have the following scenarios:

- provider/designate.go::Records() returns all DNS records

- registry/txt.go::Records() effectively returns EndPoints, so TXT records related to an EndPoint are removed

- registry/txt.go::ApplyChanges() regenerates the TXT records from scratch, but the ProviderSpecific and Labels come from the associated EndPoint rather than the TXT records, and so the unique identifier for the TXT record is lost

Hence the need to call provider/designate.go::Records() in order to obtain the metadata from the original TXT record.

Now interestingly enough I see that in registry/txt.go::Records() I see this:

```
	if im.cacheInterval > 0 {
		im.recordsCache = endpoints
		im.recordsCacheRefreshTime = time.Now()
	}
```

It seems to me that there is scope for caching the original Provider.Records() information, at least for a single run of the controller which makes sense from a consistency point of view.

I'm not sure whether the cache should be maintained per provider, or at a higher level by the registry though - I don't have enough experience with other Providers to know whether this is something that is globally useful, or a specific quirk of the txt registry.

Reply:
@mcayland If I got your third point right:
* The provider specifics of the original endpoint are copied to the TXT endpoint [here](https://github.com/kubernetes-sigs/external-dns/blob/master/registry/txt.go#L148).
* In the `Records()` function of the designate provider you could match original and TXT endpoints to get the record identifier of the TXT record and add a provider specific called `txtRecordIdentifier` or so to the original endpoint, which will then be copied to the TXT record again by the registry.

I'm not into the details of the designate provider, but I would imagine it could work this way.

Reply:
As for the problem that this would cause unneeded upsert operations, we are working on the same problem for AWS Route 53 alias records [here](https://github.com/kubernetes-sigs/external-dns/pull/1356/files#diff-953723c783e38e064b22c730117ff8b9R479-R502). We want to introduce another method `ModifyEndpoints` in the `Provider` interface where you could add the provider specifics that you also add in `Records()` to make both "states" (DNS provider vs cluster state) match before external-dns actually creates the change plan.

Reply:
Yes I think that is one way of doing this - I need to check the details as it has been a while since I last looked at this. I think the main objection was calling `Records()` again to do the comparison as part of the update?

Reply:
With the approach I described in my previous comment, there is no need to call `Records` again.